### DISCUSSION TITLE
Custom Scope Provider to propose and accept values according to the specified type

### DISCUSSION THREAD

**Discussion by theogiraudet** [eclipse-langium/langium#1244] [Q&A] [answered]
Hello,

For a personal project, I'm trying to design a small ERA (Entity-Relation-Attribute) language to describe metamodel.
In my AST, I have "Enumeration" as concept of my language and I want to allows the user to specify a default value when an attribute is typed to an enumeration. For that, I want to develop a custom scope provider in order to only propose the possible litterals of the enumeration used as type in the auto-completion.

Here is a small example of my language:

```
enumeration Axis {
  x, y, z
}

entity Rotation {
  prop degree: integer = 0
  prop axis: Axis = x // Here I only want 'x', 'y' or 'z' as possible and proposed values.
}
```

To do that, I tried to take inspiration from your guide about [class member scoping](https://langium.org/guides/scoping/class-member/) (thanks you really much for this guide) but I'm still a bit confused about how to proceed.
Here is my current getScope function:
```ts
override getScope(context: ReferenceInfo): Scope {
        if (context.property === 'literalRef') {
            console.log(context.container.$type)
            if(isEnumerationLiteralValue(context.container)) {
                const literalValue = context.container;
                if(isAttributeMember(literalValue.$container)) {
                    const attributeMember = literalValue.$container
                    if(isEnumerationType(attributeMember.type)) {
                        return this.createScopeForNodes(attributeMember.type.enumerationRef.ref!.literals);
                    }

                }
            }
        }
        return super.getScope(context);
    }
```

And here is a fragment of my interfaces generated by Langium:
```ts
interface Entity {
    name: string
    members: Member[]
}

interface Enumeration {
    name: string
    literals: EnumerationLiteral[]
}

interface EnumerationLiteral {
    name: string
}

type Member = …. | AttributeMember

interface AttributeMember {
    name: string
    type: AttributeType
    value?: Value
}

type AttributeType = ... | EnumerationType 

interface EnumerationType {
    enumerationRef: @Enumeration
}

type Value = ... | EnumerationLiteralValue

interface EnumerationLiteralValue {
    literalRef: @EnumerationLiteral
}
```
The function works almost the way I want it to, the correct values are accepted and the goto works as expected. However, when I put my cursor just after the `=` sign and I trigger the auto-completion, my values aren't proposed.
I tried to check why with the debugger and encountered a behavior I don't really understand:
![image](https://github.com/eclipse-langium/langium/assets/28688370/47c8a2d2-f799-489a-9791-6c9b94566f63)
My EnumerationLiteralValue seems to be a child of my Entity even if this isn't a valid structure according to my types.
When I check the children of my Entity, my EnumerationLiteralValue isn't in it.

Can you give me some advice on how to achieve my goals?
The complet code: https://github.com/theogiraudet/ts-metamodel
Thanks!

**Comment by cdietrich** [comment, marked as answer]
maybe you are also running into https://github.com/eclipse-langium/langium/issues/1216,
you can try if backporting https://github.com/eclipse-langium/langium/pull/1239 solves your issue

**Reply by theogiraudet** [reply]
Thank you very much, it was the same issue and this one is resolved with the new Langium version.
